File: ts/earthbucks-lib/src/iso-buf-writer.ts
import { SysBuf } from "./iso-buf.js";
import { U8, U16, U32, U64 } from "./numbers.js";

export class IsoBufWriter {
  bufs: SysBuf[];

  constructor(bufs?: SysBuf[]) {
    this.bufs = bufs ? bufs.map((arr) => SysBuf.from(arr)) : [];
  }

  getLength(): number {
    let len = 0;
    for (const buf of this.bufs) {
      len += buf.length;
    }
    return len;
  }

  toIsoBuf(): SysBuf {
    return SysBuf.concat(this.bufs);
  }

  write(buf: SysBuf): this {
    this.bufs.push(buf);
    return this;
  }

  writeU8(u8: U8): this {
    const buf = SysBuf.alloc(1);
    buf.writeUInt8(u8.n, 0);
    this.write(buf);
    return this;
  }

  writeU16BE(u16: U16): this {
    const buf = SysBuf.alloc(2);
    buf.writeUInt16BE(u16.n, 0);
    this.write(buf);
    return this;
  }

  writeU32BE(u32: U32): this {
    const buf = SysBuf.alloc(4);
    buf.writeUInt32BE(u32.n, 0);
    this.write(buf);
    return this;
  }

  writeU64BE(u64: U64): this {
    const buf = SysBuf.alloc(8);
    buf.writeBigInt64BE(u64.bn);
    this.write(buf);
    return this;
  }

  writeVarInt(u64: U64): this {
    const buf = IsoBufWriter.varIntBuf(u64);
    this.write(buf);
    return this;
  }

  static varIntBuf(bn: U64): SysBuf {
    let buf: SysBuf;
    const n = bn.n;
    if (n < 253) {
      buf = SysBuf.alloc(1);
      buf.writeUInt8(n, 0);
    } else if (n < 0x10000) {
      buf = SysBuf.alloc(1 + 2);
      buf.writeUInt8(253, 0);
      buf.writeUInt16BE(n, 1);
    } else if (n < 0x100000000) {
      buf = SysBuf.alloc(1 + 4);
      buf.writeUInt8(254, 0);
      buf.writeUInt32BE(n, 1);
    } else {
      buf = SysBuf.alloc(1 + 8);
      buf.writeUInt8(255, 0);
      buf.writeBigInt64BE(bn.bn, 1);
    }
    return buf;
  }
}

File: rs/earthbucks_lib/src/iso_buf_writer.rs
use byteorder::{BigEndian, WriteBytesExt};

// add Default
#[derive(Default)]
pub struct IsoBufWriter {
    pub bufs: Vec<Vec<u8>>,
}

impl IsoBufWriter {
    pub fn new() -> IsoBufWriter {
        IsoBufWriter { bufs: Vec::new() }
    }

    pub fn with_buffers(buffers: Vec<Vec<u8>>) -> IsoBufWriter {
        IsoBufWriter { bufs: buffers }
    }

    pub fn get_length(&self) -> usize {
        self.bufs.iter().map(|buf| buf.len()).sum()
    }

    pub fn to_iso_buf(&self) -> Vec<u8> {
        let mut result = Vec::new();
        for buf in &self.bufs {
            result.extend(buf);
        }
        result
    }

    pub fn write_iso_buf(&mut self, buf: Vec<u8>) {
        self.bufs.push(buf);
    }

    pub fn write_reverse(&mut self, buf: Vec<u8>) {
        let mut buf2 = vec![0; buf.len()];
        for (i, &item) in buf.iter().rev().enumerate() {
            buf2[i] = item;
        }
        self.bufs.push(buf2);
    }

    pub fn write_u8(&mut self, n: u8) {
        self.bufs.push(vec![n]);
    }

    pub fn write_u16_be(&mut self, n: u16) {
        let mut buf = vec![0; 2];
        buf.as_mut_slice().write_u16::<BigEndian>(n).unwrap();
        self.bufs.push(buf);
    }

    pub fn write_u32_be(&mut self, n: u32) {
        let mut buf = vec![0; 4];
        buf.as_mut_slice().write_u32::<BigEndian>(n).unwrap();
        self.bufs.push(buf);
    }

    pub fn write_u64_be(&mut self, n: u64) {
        let mut buf = vec![0; 8];
        buf.as_mut_slice().write_u64::<BigEndian>(n).unwrap();
        self.bufs.push(buf);
    }

    pub fn var_int_buf(n: u64) -> Vec<u8> {
        let mut buf = vec![];
        if n < 253 {
            buf.push(n as u8);
        } else if n < 0x10000 {
            buf.push(253);
            buf.write_u16::<BigEndian>(n as u16).unwrap();
        } else if n < 0x100000000 {
            buf.push(254);
            buf.write_u32::<BigEndian>(n as u32).unwrap();
        } else {
            buf.push(255);
            buf.write_u64::<BigEndian>(n).unwrap();
        }
        buf
    }

    pub fn write_var_int(&mut self, n: u64) -> &mut Self {
        let buf = IsoBufWriter::var_int_buf(n);
        self.write_iso_buf(buf);
        self
    }
}
